src/lib/parserRaw.c

```diff
@@ -583,6 +583,119 @@ RdbStatus elementRawSet(RdbParser *p) {
     }
 }
 
+RdbStatus elementRawModule(RdbParser *p) {
+    typedef enum RAW_MODULE_STATES {
+        ST_RAW_MODULE_GET_ID=0,
+
+        /* Following enums are aligned to module-opcodes. Static assert below. */
+        ST_RAW_MODULE_OPCODE_SINT,
+        ST_RAW_MODULE_OPCODE_UINT,
+        ST_RAW_MODULE_OPCODE_FLOAT,
+        ST_RAW_MODULE_OPCODE_DOUBLE,
+        ST_RAW_MODULE_OPCODE_STRING_CALL_STR,
+
+        ST_RAW_MODULE_OPCODE_STRING_STR_RETURN, /* return from sub-element string parsing */
```

Comment:
i don't understand (yet) this mechanism of CALL_STR and STR_RETURN.
maybe it's explained in code i didn't yet get to.

Reply:
Available for a while and exaplained in the code next to function `subElementCall` and struct `ParsingSubElement` :
```
/****************************************************************
 * Sub-Element parsing
 *
 * The parser can handle one level of nested parsing-elements (PE), whereby a PE
 * may be called by another PE and control is returned to the caller once the
 * parsing of sub-element is complete.
 *
 * Currently, this functionality is only adapted and utilized by raw string element
 * which can also run as sub-element of other data-types. It takes care to write the
 * processed string to raw-aggregator, as part of RESTORE of current command.
 *
 * In addition it returns the string decoded by allocating bulkUnmanagedAlloc()
 * (The caller won't need to release returned data yet restrictd to use it within
 * its current parsing-element state).
 ****************************************************************/
```

Reply:
so for module data types, you this sub-element for strings, and but not for other primitives?
is that described somewhere?
i understand it let's you break big module payloads to many individual calls (even split the string to several calls and avoid blocking IO), right?
what would happen if a module payload is composed of a ton of integers (let's say 100mb worth of integers)?

Reply:
Right.

It will call inefficiently to string-element multiple times through `parserMainLoop()`. Those inefficient calls can be optimized by making direct call and avoid from passing through `parserMainLoop()`. For example, instead of : 
```
            case ST_RAW_MODULE_OPCODE_STRING_CALL_STR: {
                /* call raw string as subelement */
                return subElementCall(p, PE_RAW_STRING, ST_RAW_MODULE_OPCODE_STRING_STR_RETURN);
```

We shall have:
```
            case ST_RAW_MODULE_OPCODE_STRING_CALL_STR: {
                /* call raw string as subelement */
                subElementCall(p, PE_RAW_STRING, ST_RAW_MODULE_OPCODE_STRING_STR_RETURN);
                IF_NOT_OK_RETURN(elementRawString(p)); 
                break;   /* short transition between module states without passing through main-loop */

```

I should reach later to performance phase. Yet to plan what is goanna be tested.

Reply:
Maybe I missed your point. If it is native module integer then it won't call to string and the transition between the states of the module will be made only within element-module boundary (i.e. function `elementRawModule()`)

Reply:
my last question was if it'll break out in between module integers and return to the caller?
and for string we use the sub-element in order to break if further (i.e. break the string into smaller chunks)?

Reply:
Each new (sub)element+state is being updated and can reinvoked in case of early return. 
Most of the librdb tests run also in a mode that fails all "new" read and simulates RDB_STATUS_WAIT_MORE_DATA from RDB reader/buffer, which enforces the parser to return from its current "state" and reinvoke again.